File: leetcode/SumRoottoLeafNumbers/sumNumbers.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    int sum;
    public int sumNumbers(TreeNode root) {
        // Start typing your Java solution below
        // DO NOT write main() function
        sum = 0;
        helper(root, "");
        return sum;
    }
    private void helper(TreeNode root, String path) {
        if (root == null) return;
        path += root.val;
        if (root.left == null && root.right == null) {
            sum += Integer.parseInt(path);
        }
        helper(root.left, path);
        helper(root.right, path);
    }
}

File: leetcode/SumRoottoLeafNumbers/sumNumbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* root, int cur, int& ans) {
        if (!root) return;
        cur = cur * 10 + root->val;
        if (!root->left && !root->right) ans += cur;
        dfs(root->left, cur, ans);
        dfs(root->right, cur, ans);
    }
    int sumNumbers(TreeNode* root) {
        int ans = 0, cur = 0;
        dfs(root, cur, ans);
        return ans;
    }
};
